### PR TITLE
HPCC-17865 Provide documentation on recent changes to roxiemem

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln please review.

File: roxie/roxiemem/DOCUMENTATION.rst [hunk 1]
```diff
@@ -86,12 +86,50 @@ Allocations in the huge heap can be expanded and shrunk using the resizeRow() fu
...
+The scanning heaps have better best-case performance, but worse worse-case performance (if large numbers of rows
+need to be scanned before a free row is found).  The best-case tends to be true if only one thread/activity is
+accessing a particular heap, and the worse-case if multiple activities are accessing a heap, particularly if the rows
+are being buffered.  It is the default for thor which tends to have few active allocators, but not for roxie, which
+tends to have large numbers.
+
+Delay Release
+-------------
+This is another varation on the scanning allocator, which further reduces the number of atomic operations.  Usually
+when a row is allocated the link count on the heaplet is increased, and when it is free the link count is
```
Comment (mckellyln) [inline roxie/roxiemem/DOCUMENTATION.rst]: when it is freed (past tense) to match allocated tense.

File: roxie/roxiemem/DOCUMENTATION.rst [hunk 2]
```diff
@@ -86,12 +86,50 @@ Allocations in the huge heap can be expanded and shrunk using the resizeRow() fu
...
+an atomic operation for each allocation and for each free.  The scanning allocator uses an alternative approach.  When
+a row is freed the row header is marked, and a row is allocated by scanning through the heaplet for rows that have
+been marked as free.  Scanning uses atomic store and get, rather than more expensive synchronized atomic operations,
+so is generally faster than the linked list - provided a free row is found fairly quickly.
+
+The scanning heaps have better best-case performance, but worse worse-case performance (if large numbers of rows
+need to be scanned before a free row is found).  The best-case tends to be true if only one thread/activity is
+accessing a particular heap, and the worse-case if multiple activities are accessing a heap, particularly if the rows
+are being buffered.  It is the default for thor which tends to have few active allocators, but not for roxie, which
+tends to have large numbers.
```
Comment (mckellyln) [inline roxie/roxiemem/DOCUMENTATION.rst]: large numbers of allocators.

**Review by mckellyln — APPROVED** [review]
minor wording perhaps

**Comment by mckellyln** [comment]
@ghalliday looks good, just some minor comments on a few words.

**Comment by ghalliday** [comment]
@richardkchapman review changes made and commits squashed

**Pull request merged**